Return container command exit codes in kubectl run/exec

sttts opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/26424
Based on https://github.com/kubernetes/kubernetes/pull/25273.

TODO:
- [x] add e2e tests
- [x] investigate `kubectl run` exit code for `--restart=Never` (compare issue #24533 and PR #25253)
- [x] document exit codes

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/26541)

<!-- Reviewable:end -->

sttts commented:
@ncdc this is ready for a first review. Review commit-by-commit would be easiest probably.

ncdc commented on pkg/client/unversioned/remotecommand/errors.go:
```diff
@@ -0,0 +1,22 @@
+
+package remotecommand
+
+// errorDecoder interprets the data on the error channel and create a go error object from it.
```
and creates

ncdc commented on pkg/client/unversioned/remotecommand/v3.go:
```diff
@@ -108,3 +110,34 @@ func (e *streamProtocolV3) stream(conn streamCreator) error {
+	return e.ExitError.Exited
+}
+
+func (e *decodedExitError) ExitStatus() int {
```
Call this ExitCode?

sttts replied:
It's part of the `util.exec.ExitError` interface we are implementing here. I have added a line to make this explicit.

sttts replied:
Seems like ExitCode and ExitStatus are used interchangeable in the code base.

ncdc commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl/resource"
 	"k8s.io/kubernetes/pkg/runtime"
 	utilerrors "k8s.io/kubernetes/pkg/util/errors"
+	uexec "k8s.io/kubernetes/pkg/util/exec"
```
I would import as utilexec which matches the previous import

sttts replied:
fixed

sttts replied:
done

ncdc commented on pkg/kubelet/server/remotecommand/exec.go:
```diff
@@ -25,10 +26,23 @@ import (
+	Message string `json:"message"`
+
+	// ExitStatus is the process exec code.
+	ExitStatus int `json:"exitstatus"`
```
I'd call this ExitCode.

sttts replied:
done

ncdc commented on pkg/client/unversioned/remotecommand/v3.go:
```diff
@@ -108,3 +110,34 @@ func (e *streamProtocolV3) stream(conn streamCreator) error {
 	return <-errorChan
 }
+
+type decodedExitError struct {
```
Why do we need this type?

ingvagabund replied:
It is used in kubectl e2e test to perform type conversion.

ncdc replied:
@ingvagabund I don't see that usage.

@sttts could you alternatively get rid of `decodedExitError` and instead make `remotecommand.ExitError` implement `uexec.ExitError`?

sttts replied:
done

ncdc commented on pkg/kubectl/cmd/run.go:
```diff
@@ -325,6 +370,21 @@ func waitForPodRunning(c *client.Client, pod *api.Pod, out io.Writer) (status ap
 
+func waitForPodTerminated(c *client.Client, ns, name string, out io.Writer) (pod *api.Pod, err error) {
+	for {
+		pod, err := c.Pods(ns).Get(name)
```
What about a Watch instead of Get+Sleep?

sttts replied:
done

dims commented:
@sttts This can go in now. right?

sttts commented:
@dims it's only waiting for some review love

ncdc commented:
I plan to review today

On Wednesday, July 20, 2016, Dr. Stefan Schimanski <notifications@github.com>
wrote:

> @dims <https://github.com/dims> it's only waiting for some review love
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541#issuecomment-233934065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYk0YNGaxkhCYO0UR0LdtAz3zDTKYks5qXhPxgaJpZM4Ip8Gm>
> .
>

dims commented:
@sttts We seem to be still printing a string. Can we get rid of that please?

```
[dims@ubuntu-brix 11:13] ~ ⟩ kubectl exec -it sleep-548116132-o1mkp -- /bin/sh -c "exit 5"
Error executing in Docker Container: 5
[dims@ubuntu-brix 11:13] ~ ⟩ echo $?
5
```

ncdc commented:
@dims this is still on my list of things to review. It didn't happen yesterday unfortunately :-(

dims commented:
@sttts - Found one more unexpected thing. Don't know if this was old behavior or new.. I ran a python command to test output to stdout, stderr and the exit code as follows.

In addition to `Error executing in Docker Container: 5` shows up in stderr (which we talked about earlier), the `writing to stderr` text is written to stdout (NOT stderr as i expected).

Here's my kubectl command line:
```
[dims@ubuntu-brix 17:43] ~/kubernetes ⟩ kubectl exec -it sleep222-1848484887-fptg1 -- python -c 'import sys;sys.stdout.write("writing to stdout\n");sys.stderr.write("writing to stderr\n");sys.exit(5)'
writing to stdout
writing to stderr
Error executing in Docker Container: 5
```

Here's my setup for testing this:
```
[dims@ubuntu-brix 17:43] ~/kubernetes ⟩ kubectl run --image=gcr.io/google-containers/python:v1 sleep222 -- /bin/sh -c "sleep 1000"
deployment "sleep222" created
[dims@ubuntu-brix 17:46] ~/kubernetes ⟩ kubectl get po
NAME                        READY     STATUS    RESTARTS   AGE
sleep222-1848484887-fptg1   0/1       Pending   0          7s
```

ncdc commented:
@sttts I'm not going to be able to review this for a few days. @ingvagabund would you have time to do it?

@dims if you exec with a tty (`-t`), anything you write in the container to stderr will be written to stderr within the container, but what you get back to your local terminal's stdout is the combination of both stdout and stderr from the container because you're using a tty. If you want to redirect stderr in the container to something else, you would need to do that e.g. as part of a bash invocation that's part of your exec call. If you want to be able to control the redirection from your local terminal, you would need to do something like (super pseudo) `kubectl exec -i mypod -- stdout.print('out'), stderr.print('err') 1>stdout.log 2>stderr.log`

sttts commented:
@dims @ncdc I have removed the error output on ExitCode errors now.

I can confirm @ncdc's explanation about tty and stderr behavior: we behave like docker-run:

```shell
$ docker run -it busybox /bin/sh -c "echo foo 1>&2" 2>/dev/null
foo
$ docker run -i busybox /bin/sh -c "echo foo 1>&2" 2>/dev/null
$ docker run -i busybox /bin/sh -c "echo foo 1>&2"
foo

$ kubectl exec -it ctrlc -- /bin/sh -c "echo foo 1>&2" 2>/dev/null
foo
$ kubectl exec -i ctrlc -- /bin/sh -c "echo foo 1>&2" 2>/dev/null
$ kubectl exec -i ctrlc -- /bin/sh -c "echo foo 1>&2"
foo
```

ingvagabund commented:
@ncdc taking my chances

dims commented:
@sttts thanks!

ingvagabund commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -84,19 +86,21 @@ func DefaultBehaviorOnFatal() {
-		msg += "\n"
-	}
+func fatal(msg string, code int) {
+	if len(msg) > 0 {
```
Great. It does not make sense to print empty line.

ingvagabund commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -204,25 +206,28 @@ func StandardErrorMessage(err error) (string, bool) {
+// lines. The returned string will end with a newline. The prefix should not end in . or : or \n.
 func MultilineError(prefix string, err error) string {
+	if err == nil {
+		return fmt.Sprintf("%s.\n", prefix)
```
Does it make sense to print the prefix if there is no error?

sttts replied:
In fact it's pretty strange to call MultilineError with a nil error. I will check whether this can be cleaned up easily.

sttts replied:
In fact, what the function actually means is this:

```
func multilineErrorString(err string, subErr error) string {
```

Then the check for nil actually makes sense and it's fine to print the `err` when subErr is nil.

ingvagabund commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -204,25 +206,28 @@ func StandardErrorMessage(err error) (string, bool) {
+			return fmt.Sprintf("%s: %v\n", prefix, err)
 		case 1:
-			return fmt.Sprintf("%s%v\n", prefix, messageForError(errs[0]))
+			return fmt.Sprintf("%s: %v\n", prefix, messageForError(errs[0]))
```
This makes the error message more transparent. Nice.

ingvagabund commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -113,48 +117,46 @@ func checkErrWithPrefix(prefix string, err error) {
-		case len(noMatch.PartialResource.Version) > 0:
-			handleErr(fmt.Sprintf("%sthe server doesn't have a resource type %q in version %q", pref, noMatch.PartialResource.Resource, noMatch.PartialResource.Version))
+		handleErr(MultilineError(prefix, errs), DefaultErrorExitCode)
+	case clientcmd.IsConfigurationInvalid(err):
```
If ``clientcmd.IsConfigurationInvalid(err)`` and ``noMatch, ok := err.(*meta.NoResourceMatchError); ok`` can hold at the same time, order of testing should not be switched.

sttts replied:
I guess we can assume this here.

ingvagabund commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -113,48 +117,46 @@ func checkErrWithPrefix(prefix string, err error) {
-		msg = err.Error()
-		if !strings.HasPrefix(msg, "error: ") {
-			msg = fmt.Sprintf("error: %s", msg)
+			msg, ok := StandardErrorMessage(err)
```
Based on [1], if the ``default`` branch is picked, ``err := err.(type)`` is identity assignment, i.e. ``err = err``. So it make sense to make this the default. Though, this behaviour is not obvious to me. Can you comment the line with somethine like "err := err.(type) equals to err = err if the default branch is picked" or similar?

[1] https://golang.org/doc/effective_go.html#type_switch

sttts replied:
done

ingvagabund commented on pkg/kubectl/cmd/util/helpers_test.go:
```diff
@@ -213,72 +214,83 @@ func (f *fileHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	testCheckError(t, []checkErrTestCase{
 		{
 			errors.NewInvalid(api.Kind("Invalid1"), "invalidation", field.ErrorList{field.Invalid(field.NewPath("field"), "single", "details")}),
-			`Error from server: Invalid1 "invalidation" is invalid: field: Invalid value: "single": details`,
```
"Error from server" suddenly disappeared, why?

sttts replied:
Because the testcase was wrong, testing a code path that could never happen:

`checkErr` calls `handleErr` a number of times. In practice this quits the program. In the test though it does not, i.e. the last `handleErr` wins, giving the `Error from server:` prefix.

I fixed this behavior and adapted the test to reality.

ingvagabund commented on pkg/kubectl/cmd/util/helpers_test.go:
```diff
@@ -213,72 +214,83 @@ func (f *fileHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	})
+}
+
+func TestCheckExitError(t *testing.T) {
```
Looks good, testing the new exit code parameter.

ingvagabund commented on pkg/kubectl/cmd/run.go:
```diff
@@ -304,45 +350,98 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
 }
 
-func waitForPodRunning(c *client.Client, pod *api.Pod, out io.Writer, quiet bool) (status api.PodPhase, err error) {
+func waitForPod(c *client.Client, ns, name string, exitCondition func(*api.Pod) bool, tick func(*api.Pod)) (*api.Pod, error) {
```
``tick`` argument together with semantic of ``waitForPod`` awakens interpretation of ``tick`` to be the ticker function controlling the for loop inside. What about to change ``tick`` to ``notify``?

sttts replied:
`tick` comes from the Go `ticker`: https://golang.org/pkg/time/#Ticker. It's called on each tick of the timer.

ingvagabund commented on pkg/kubectl/cmd/run.go:
```diff
@@ -274,9 +286,43 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
+			if len(pod.Status.ContainerStatuses) == 0 || pod.Status.ContainerStatuses[0].State.Terminated == nil {
+				return fmt.Errorf("pod %s/%s failed with unknown exit code", pod.Namespace, pod.Name)
+			}
+			if len(pod.Status.ContainerStatuses) > 1 {
```
``kubectl run`` should create a pod with one container only. So this condition will never be true (most likely). However, as the code just prints messages and one can never be sure of non-existence of some black magic, it does not hurt to keep it there.

sttts replied:
Have remove that code.

sttts commented:
Addressed all review comments.

ingvagabund commented:
@kubernetes/kube-ui-maintainers it looks good by design, still, would be great to have someone from ui to take a look at it form the general/specific point of view.

ncdc commented:
cc @kubernetes/rh-ux

aveshagarwal commented on test/e2e/framework/util.go:
```diff
@@ -2034,7 +2036,15 @@ func (b kubectlBuilder) Exec() (string, error) {
+				rc = int(ee.Sys().(syscall.WaitStatus).ExitStatus())
+			}
+			return "", uexec.CodeExitError{
+				Err:  fmt.Errorf("Error running %v:\nCommand stdout:\n%v\nstderr:\n%v\nerror:\n%v\n", cmd, cmd.Stdout, cmd.Stderr, err),
```
nit: First letter lowercase, although I have seen mix of lower and upper cases but majority seems to be with lower case.

sttts replied:
Looks like capital case is consistently used in this file.

smarterclayton replied:
That's really a flaw in this package - the rule is all lower, with extremely limited exceptions.

aveshagarwal commented on pkg/kubectl/cmd/run.go:
```diff
@@ -254,11 +256,21 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 		}
-		err = handleAttachPod(f, client, attachablePod, opts, quiet)
+		err = handleAttachPod(f, client, attachablePod.Namespace, attachablePod.Name, opts, quiet)
 		if err != nil {
```
nit: Wonder, we are using quiet here but we are reading that in the first line of this function?

sttts replied:
moved down.

sttts replied:
Moved up again as it is needed above in the latest master.

aveshagarwal commented on pkg/kubectl/cmd/run.go:
```diff
@@ -276,9 +288,35 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
+		if !waitForExitCode {
+			return nil
+		}
+
```
wonder why this block is here and not before "if remove" block?

ncdc replied:
Because you have to perform the removal no matter what. If you move this switch block up, it would return too early and the remove wouldn't happen.

ncdc commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -165,13 +167,13 @@ func statusCausesToAggrError(scs []unversioned.StatusCause) utilerrors.Aggregate
 // This method is generic to the command in use and may be used by non-Kubectl
 // commands.
-func StandardErrorMessage(err error) (string, bool) {
+func standardErrorMessage(err error) (string, bool) {
```
Don't unexport - you never know who might be using it downstream (OpenShift)

sttts replied:
done

ncdc commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -203,31 +205,34 @@ func StandardErrorMessage(err error) (string, bool) {
-	if agg, ok := err.(utilerrors.Aggregate); ok {
+// multilineErrorString returns a string representing an error that splits sub errors into their own
+// lines. The returned string will end with a newline. The prefix should not end in . or : or \n.
+func multilineErrorString(err string, subErr error) string {
```
Don't unexport - you never know who might be using it downstream (OpenShift)

sttts replied:
Origin at least doesn't.

ncdc commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -203,31 +205,34 @@ func StandardErrorMessage(err error) (string, bool) {
+// multipleErrors returns a newline delimited string containing
 // the prefix and referenced errors in standard form.
-func MultipleErrors(prefix string, errs []error) string {
+func multipleErrors(prefix string, errs []error) string {
```
Don't unexport - you never know who might be using it downstream (OpenShift)

sttts replied:
This in fact is used downstream. Having invisible downstream usage of those auxiliary functions really hurts.

ncdc commented on pkg/kubelet/server/remotecommand/exec.go:
```diff
@@ -25,10 +26,23 @@ import (
+	// Message is the actual human readable text error message.
+	Message string `json:"message"`
+
+	// ExitCode is the process exec code.
```
s/exec/exit/

ncdc commented on pkg/kubelet/server/remotecommand/exec.go:
```diff
@@ -51,8 +65,16 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
+			m.Message = fmt.Sprintf("error executing command in container: %v", err)
+		}
+		runtime.HandleError(errors.New(m.Message))
+		bs, _ := json.Marshal(&m)
```
Why not check the error here?

sttts replied:
not really much we can do here in case of an error. Added a return on err for uniformity.

sttts commented:
@aveshagarwal @ncdc addressed all comments.

sttts commented:
@k8s-bot test this issue: #26760

ncdc commented on pkg/kubectl/cmd/run.go:
```diff
@@ -306,45 +345,100 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
+
+	tick(pod)
+
+	w, err := c.RESTClient.Get().
```
Could you use `w, err := c.Pods(ns).Watch(api.ListOptions{ResourceVersion: pod.ResourceVersion})` instead?

sttts replied:
done

ncdc commented on pkg/kubectl/cmd/run.go:
```diff
@@ -306,45 +345,100 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
-		pod, err := c.Pods(pod.Namespace).Get(pod.Name)
-		if err != nil {
-			return api.PodUnknown, err
+		select {
```
If you didn't have a ticker, I would say use `kubectl.WatchLoop`

sttts replied:
With some channel+go-routine magic we could use a watch. Will take a look.

smarterclayton replied:
This should be using `watch.Wait`, since you're effectively doing the same thing.  See all the client conditions for this, and add the proper ones there or reuse the existing.

sttts replied:
Not sure which `watch.Wait` you mean. We have `watch.Until`. But `kubectl.WatchLoop` with signal support looks like a better fit.

smarterclayton replied:
WatchLoop should be using `watch.Until`.  It should be layered.

smarterclayton replied:
Also `WatchLoop` should be using the `interrupt` package, not listening to notify itself, and support being called in a parent interrupt context.

ncdc commented:
A few comments. Everything else looks good. I would like @kubernetes/kubectl and @kubernetes/rh-ux to review the changes to the files under `pkg/kubectl` if possible.

fabianofranz commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -203,26 +205,29 @@ func StandardErrorMessage(err error) (string, bool) {
+// lines. The returned string will end with a newline. The prefix should not end in . or : or \n.
+func MultilineErrorString(err string, subErr error) string {
+	if subErr == nil {
+		return fmt.Sprintf("%s.\n", err)
```
I don't recommend making assumptions/requirements about the string you get here. This is used through `CheckErr` with messages that come from various different places (API, commands, etc) and it is quite unpredictable. So either remove the `.` or make it smarter (trim spaces then add the dot if not present in the end, etc).

sttts replied:
done

fabianofranz commented:
One comment, otherwise LGTM from a `kubectl` perspective.

smarterclayton commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -203,26 +205,29 @@ func StandardErrorMessage(err error) (string, bool) {
-// lines. The returned string will end with a newline.
-func MultilineError(prefix string, err error) string {
-	if agg, ok := err.(utilerrors.Aggregate); ok {
+// MultilineErrorString returns a string representing an error that splits sub errors into their own
```
Why are you changing the signature here?  MultilineError means something really specific, and changing this behavior is really defining a new method.

sttts replied:
There was a case (one out of two) where it was called with a nil err, giving strange output. I have reverted the change here and added a clause to avoid the nil on the callers side.

sttts replied:
@smarterclayton please also review `TestCheckInvalidErr`. The old test cases were not testing reality, now they do.

smarterclayton commented on pkg/kubelet/server/remotecommand/exec.go:
```diff
@@ -51,8 +85,19 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
-		msg := fmt.Sprintf("error executing command in container: %v", err)
-		runtime.HandleError(errors.New(msg))
-		fmt.Fprint(ctx.errorStream, msg)
+		m := ExitError{}
```
We should not be creating a new error type, when we already have one (`APIStatus`).  That's what we should be returning here, and we need to obey versioning rules w.r.t. the error (define in the protocol what sort of error we return).

I'm also concerned that this change is not backwards compatible, in that old clients will get a new, unexpected value on the error stream.

ncdc replied:
Yeah, I think an old client/new server combo would be busted.

smarterclayton replied:
K, new protocol!

On Wed, Aug 3, 2016 at 3:48 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/kubelet/server/remotecommand/exec.go
> <https://github.com/kubernetes/kubernetes/pull/26541#discussion_r73407352>
> :
>
> > @@ -51,8 +85,19 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
> >
> >  	err := executor.ExecInContainer(podName, uid, container, cmd, ctx.stdinStream, ctx.stdoutStream, ctx.stderrStream, ctx.tty, ctx.resizeChan)
> >  	if err != nil {
> > -		msg := fmt.Sprintf("error executing command in container: %v", err)
> > -		runtime.HandleError(errors.New(msg))
> > -		fmt.Fprint(ctx.errorStream, msg)
> > +		m := ExitError{}
>
> Yeah, I think an old client/new server combo would be busted.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541/files/0a10e2dcdaf86a90b5a718e96d44cc9e5286ca06#r73407352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-pPdIwS6AgLL7f0jCQxkMcTXP2oks5qcPCZgaJpZM4Ip8Gm>
> .
>

sttts replied:
Good point. Old clients would see the JSON and interpret that as the error text. I am on the way of refactoring this into a v4 protocol now.

In this context @ncdc: why is there no version handshake for the websocket code path?

ncdc replied:
There is but it's done differently. I'll link it when I'm at my desk.

On Thursday, August 4, 2016, Dr. Stefan Schimanski <notifications@github.com>
wrote:

> In pkg/kubelet/server/remotecommand/exec.go
> <https://github.com/kubernetes/kubernetes/pull/26541#discussion_r73508473>
> :
>
> > @@ -51,8 +85,19 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
> >
> >  	err := executor.ExecInContainer(podName, uid, container, cmd, ctx.stdinStream, ctx.stdoutStream, ctx.stderrStream, ctx.tty, ctx.resizeChan)
> >  	if err != nil {
> > -		msg := fmt.Sprintf("error executing command in container: %v", err)
> > -		runtime.HandleError(errors.New(msg))
> > -		fmt.Fprint(ctx.errorStream, msg)
> > +		m := ExitError{}
>
> Good point. Old clients would see the JSON and interpret that as the error
> text. I am on the way of refactoring this into a v4 protocol now.
>
> In this context @ncdc <https://github.com/ncdc>: why is there no version
> handshake for the websocket code path?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541/files/0a10e2dcdaf86a90b5a718e96d44cc9e5286ca06#r73508473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYuemXq4ZNSgKYwhkg-JPZJX-d2p-ks5qcdOkgaJpZM4Ip8Gm>
> .
>

ncdc replied:
See https://github.com/kubernetes/kubernetes/blob/master/pkg/util/wsstream/conn.go#L168

ncdc replied:
The v3 protocol was added after Kubernetes v1.3 was released. Assuming we get this PR merged before v1.4 is cut, we could just add this to the v3 protocol, right? @smarterclayton

sttts replied:
That was my initial rational as well. But adding v4 is cleaner and not much more work. We have to find a way to send different error messages anyway, depending on the version.

ncdc replied:
Yeah, looking at the code the way it's currently structured, it's not
immediately easy to send back different responses based on protocol version
right here. I guess we could stick the protocol on the context object as
part of createStreams.

On Thu, Aug 4, 2016 at 8:28 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In pkg/kubelet/server/remotecommand/exec.go
> <https://github.com/kubernetes/kubernetes/pull/26541#discussion_r73511874>
> :
>
> > @@ -51,8 +85,19 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
> >
> >  	err := executor.ExecInContainer(podName, uid, container, cmd, ctx.stdinStream, ctx.stdoutStream, ctx.stderrStream, ctx.tty, ctx.resizeChan)
> >  	if err != nil {
> > -		msg := fmt.Sprintf("error executing command in container: %v", err)
> > -		runtime.HandleError(errors.New(msg))
> > -		fmt.Fprint(ctx.errorStream, msg)
> > +		m := ExitError{}
>
> That was my initial rational as well. But adding v4 is cleaner and not
> much more work. We have to find a way to send different error messages
> anyway, depending on the version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541/files/0a10e2dcdaf86a90b5a718e96d44cc9e5286ca06#r73511874>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYgdC0PlazH89hhQjAeVBAHJREkwOks5qcdrSgaJpZM4Ip8Gm>
> .
>

sttts replied:
I am replacing `errorStream` with `func writeError(message string, rc *int)`. I have that for http, have to understand the websocket part now.

ncdc replied:
That works too :-)

On Thu, Aug 4, 2016 at 8:42 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In pkg/kubelet/server/remotecommand/exec.go
> <https://github.com/kubernetes/kubernetes/pull/26541#discussion_r73513635>
> :
>
> > @@ -51,8 +85,19 @@ func ServeExec(w http.ResponseWriter, req *http.Request, executor Executor, podN
> >
> >  	err := executor.ExecInContainer(podName, uid, container, cmd, ctx.stdinStream, ctx.stdoutStream, ctx.stderrStream, ctx.tty, ctx.resizeChan)
> >  	if err != nil {
> > -		msg := fmt.Sprintf("error executing command in container: %v", err)
> > -		runtime.HandleError(errors.New(msg))
> > -		fmt.Fprint(ctx.errorStream, msg)
> > +		m := ExitError{}
>
> I am replacing errorStream with func writeError(message string, rc *int).
> I have that for http, have to understand the websocket part now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541/files/0a10e2dcdaf86a90b5a718e96d44cc9e5286ca06#r73513635>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYmf5WdoWmeHew7puTchET1yR6Umwks5qcd4dgaJpZM4Ip8Gm>
> .
>

sttts replied:
done.

Commit:
Add protocol versions to pkg/util/wsstream

smarterclayton commented on pkg/util/wsstream/conn.go:
```diff
@@ -123,21 +124,28 @@ func handshake(config *websocket.Config, req *http.Request, allowed []string) er
 // duplex modes. Read and Write refer to whether the channel can be used as a Reader or Writer.
-func NewConn(channels ...ChannelType) *Conn {
+//
+// Protocols have the shape "channel.k8s.io" in the unversioned case and "channel.k8s.io/v4" in
```
`v4.channel.k8s.io` is just a new protocol.  There's no need to get fancy with version + protocol.

smarterclayton replied:
I don't want to add the complexity of version + protocol - it's just like mime type.

sttts replied:
it's less about being fancy, but about easy distinction what is handled inside the `Conn` (binary vs. b64) and what is provided from outside. Using the version as prefix with a dot is not simpler.

smarterclayton replied:
It's not simpler but it's consistent with the decision we made on protocols
previously.  I don't want to invent a new mixing scheme.  Each protocol is
independent.

On Fri, Aug 5, 2016 at 1:18 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In pkg/util/wsstream/conn.go
> <https://github.com/kubernetes/kubernetes/pull/26541#discussion_r73643040>
> :
>
> >  }
> >
> >  // NewConn creates a WebSocket connection that supports a set of channels. Channels begin each
> >  // web socket message with a single byte indicating the channel number (0-N). 255 is reserved for
> >  // future use. The channel types for each channel are passed as an array, supporting the different
> >  // duplex modes. Read and Write refer to whether the channel can be used as a Reader or Writer.
> > -func NewConn(channels ...ChannelType) *Conn {
> > +//
> > +// Protocols have the shape "channel.k8s.io" in the unversioned case and "channel.k8s.io/v4" in
>
> it's less about being fancy, but about easy distinction what is handled
> inside the Conn (binary vs. b64) and what is provided from outside. Using
> the version as prefix with a dot is not simpler.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26541/files/ac7867dd7be0fdf3a417cc7fb69cb95cfccb12b0..176f97d29edfe3228ce86ffddc957d7146c1e3bd#r73643040>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxxftQZ201lmO0PzuYbbXHKp0j6kks5qcseXgaJpZM4Ip8Gm>
> .
>

sttts replied:
I am fine with the dot notation, will change that. I guess a `protocolVersion` becomes `protocolPrefix` then. Can you comment on the `Conn` API? Is that ok as it is (with the renamings)? I would like to avoid the two protocols `channel.k8s.io` and `base64.channel.k8s.io` leaking outside of this package.

sttts replied:
done.

Commit:
Rectify kubectl error output

Commit:
Add return code support to kubectl-exec and -run

sttts commented:
Squashed.

smarterclayton commented on pkg/util/wsstream/conn.go:
```diff
@@ -123,21 +125,30 @@ func handshake(config *websocket.Config, req *http.Request, allowed []string) er
+// are provided all combinations of those and base64+raw are nogatiated, e.g. "", "v1", "v2" leads
+// to the protocols "base64.channel.k8s.io", "channel.k8s.io", "v1.base64.channel.k8s.io",
+// "v1.channel.k8s.io", "v2.base64.channel.k8s.io", "v2.channel.k8s.io".
+func NewConn(protocolPrefixes []string, channels ...ChannelType) *Conn {
```
New conn doesn't really know anything about protocols - if you want to make it protocol aware, you should be passing in something that abstracts protocols, rather than make NewConn protocol aware.

smarterclayton replied:
So the original intent of NewConn was to be protocol agnostic - as we're adding more logic to it it doesn't really belong in pkg/util anymore, and instead should be moved to a parallel structure to exec.  In that case you can continue to use abstraction, although channel types should really not be something you observe, but instead take a map of behaviors.

sttts replied:
The subprotocols are now under control of the remotecontrol package. Only a `NewDefaultChannelProtocols` is left, with the old two protocols.

smarterclayton commented on pkg/util/wsstream/conn.go:
```diff
@@ -154,6 +165,29 @@ func NewConn(channels ...ChannelType) *Conn {
+	for _, prefix := range prefixes {
+		if prefix != "" {
+			if errs := validation.IsDNS1123Subdomain(prefix); len(errs) > 0 {
+				glog.Fatalf("websocket protocol prefix %q is no DNS 1123 subdomain: %v", prefix, errs)
```
Don't use glog.Fatal anywhere outside of CLI code or a very limited set of init code, and even then sparingly

sttts replied:
Done, function is gone.

smarterclayton commented on pkg/util/wsstream/conn.go:
```diff
@@ -175,20 +209,44 @@ func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) ([]io.ReadWrite
+	if proto == b64 || proto == raw {
+		return "", proto
+	}
+	if strings.HasSuffix(proto, "."+b64) {
```
Don't do this - instead, accept a map of protocol -> behavior structs, and base your decisions on structured fields in the struct.  I.e.

```
type map[string]Behavior
type Behavior struct {
  Binary bool
  ChannelHandlers [](functions that handle channels)
}

sttts replied:
Done.

sttts commented on pkg/util/wsstream/conn.go:
```diff
@@ -117,41 +119,51 @@ func handshake(config *websocket.Config, req *http.Request, allowed []string) er
-		case IgnoreChannel:
-			conn.channels[i] = newWebsocketChannel(conn, byte(i), false, false)
-		}
+//
```
one line too much

sttts commented:
ptal

ncdc commented on pkg/util/wsstream/conn_test.go:
```diff
@@ -167,3 +174,96 @@ func TestBase64Conn(t *testing.T) {
+			conn := NewConn(supportedProtocols)
+			var selectedProtocol string
+			s, addr := newServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				selectedProtocol, _, _ = conn.Open(w, req)
```
This will be a data race when you try to read it below. You'll need to pass it back via a channel.

ncdc replied:
```
==================
WARNING: DATA RACE
Read by goroutine 20:
  k8s.io/kubernetes/pkg/util/wsstream.TestVersionedConn.func1()
      /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/conn_test.go:264 +0xaec
  k8s.io/kubernetes/pkg/util/wsstream.TestVersionedConn()
      /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/conn_test.go:267 +0x172
  testing.tRunner()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/testing/testing.go:473 +0xdc

Previous write by goroutine 25:
  k8s.io/kubernetes/pkg/util/wsstream.TestVersionedConn.func1.1()
      /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/conn_test.go:241 +0x80
  net/http.HandlerFunc.ServeHTTP()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/net/http/server.go:1618 +0x47
  net/http.serverHandler.ServeHTTP()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/net/http/server.go:2081 +0x206
  net/http.(*conn).serve()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/net/http/server.go:1472 +0x1565

Goroutine 20 (running) created at:
  testing.RunTests()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/testing/testing.go:515 +0x11d
  main.main()
      k8s.io/kubernetes/pkg/util/wsstream/_test/_testmain.go:74 +0x210

Goroutine 25 (finished) created at:
  net/http.(*Server).Serve()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/net/http/server.go:2137 +0x4d1
  net/http/httptest.(*Server).goServe.func1()
      /home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/net/http/httptest/server.go:237 +0xac
==================
E0812 13:44:30.040001    2134 runtime.go:64] Observed a panic: "bad read" (bad read)
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:70
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:63
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:49
/home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/runtime/asm_amd64.s:472
/home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/runtime/panic.go:443
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/stream_test.go:221
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/stream.go:131
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/stream.go:112
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/stream.go:100
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/websocket/server.go:89
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/websocket/server.go:70
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wsstream/stream.go:100
/home/vagrant/.gimme/versions/go1.6.3.linux.amd64/src/runtime/asm_amd64.s:1998
--- FAIL: TestStreamClosedDuringRead (0.00s)
        stream_test.go:199: websocket.Dial ws://127.0.0.1:45887: bad status
FAIL
FAIL    k8s.io/kubernetes/pkg/util/wsstream     0.042s
```

sttts replied:
good catch. Will put in some synchronization.

sttts replied:
Done.

sttts commented:
Rebased and fixed another gofmt issue.

pmorie commented on pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -45,8 +45,8 @@ import (
-	cmdutil.BehaviorOnFatal(func(str string) {
-		t.Errorf("Error running command: %s", str)
+	cmdutil.BehaviorOnFatal(func(str string, code int) {
+		t.Errorf("Error running command (rc=%d): %s", code, str)
```
nit, might want to make this more obvious than `rc`

sttts replied:
done

pmorie commented on pkg/kubectl/cmd/run.go:
```diff
@@ -114,7 +116,7 @@ func addRunFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolP("stdin", "i", false, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
 	cmd.Flags().BoolP("tty", "t", false, "Allocated a TTY for each container in the pod.")
-	cmd.Flags().Bool("attach", false, "If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.  Default false, unless '-i/--stdin' is set, in which case the default is true.")
+	cmd.Flags().Bool("attach", false, "If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.  Default false, unless '-i/--stdin' is set, in which case the default is true. With '--restart=Never' the image command exit code is returned.")
```
Is `Image` the right word here?  Pod process, perhaps?

sttts replied:
yes, sounds better.

pmorie commented on pkg/client/unversioned/remotecommand/errorstream.go:
```diff
@@ -0,0 +1,51 @@
+	decode(message []byte) error
+}
+
+func watchErrorStream(errorStream io.Reader, d errorStreamDecoder) chan error {
```
godoc

sttts replied:
done

pmorie commented on pkg/client/unversioned/remotecommand/errorstream.go:
```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2016

sttts replied:
Coding in the past.

pmorie commented on pkg/client/unversioned/remotecommand/v2.go:
```diff
@@ -207,3 +186,9 @@ func (p *streamProtocolV2) stream(conn streamCreator) error {
 	return <-errorChan
 }
+
+type errorDecoderV2 struct{}
```
nice to have: godoc explaining the variance between the different error decoders

sttts replied:
done

pmorie commented on pkg/client/unversioned/remotecommand/v4.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	// now that all the streams have been created, proceed with reading & copying
+
```
dead line

sttts replied:
Copy and paste from V3. Here the dead line actually makes sense as it describes the flow of the whole func, not only the next block.

pmorie commented on pkg/client/unversioned/remotecommand/v4.go:
```diff
@@ -0,0 +1,91 @@
+	return <-errorChan
+}
+
+// errorDecoderV4 interprets the json data on the error channel and creates an uexec.ExitError from it.
```
typo? `uexec`

pmorie replied:
I see now that you alias that package, but in another file -- it's not aliased in this file, which threw me off.

pmorie commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -113,48 +117,50 @@ func checkErrWithPrefix(prefix string, err error) {
 
-func checkErr(pref string, err error, handleErr func(string)) {
-	if err == nil {
+func checkErr(prefix string, err error, handleErr func(string, int)) {
```
godoc

pmorie commented on pkg/kubectl/cmd/util/helpers_test.go:
```diff
@@ -213,72 +214,83 @@ func (f *fileHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+			DefaultErrorExitCode,
 		},
 		{
 			errors.NewInvalid(api.Kind("Invalid2"), "invalidation", field.ErrorList{field.Invalid(field.NewPath("field1"), "multi1", "details"), field.Invalid(field.NewPath("field2"), "multi2", "details")}),
```
consider wrapping this mega-line

pmorie commented on pkg/kubelet/server/remotecommand/httpstream.go:
```diff
@@ -375,3 +429,34 @@ func handleResizeEvents(stream io.Reader, channel chan<- term.Size) {
+// V4ExitErrorMessage objects are marshaled as json in the error channel.
+type V4ExitErrorMessage struct {
+	// Message is the actual human readable text error message.
+	Message string `json:"message"`
```
is this always going to be populated?

sttts replied:
yes

smarterclayton replied:
Since this is in an API object this needs to be `api.Status`.

smarterclayton replied:
And the fields should be set appropriately - we should have a way to send back exit code, probably as a detail

pmorie commented on pkg/client/unversioned/remotecommand/v2.go:
```diff
@@ -207,3 +186,10 @@ func (p *streamProtocolV2) stream(conn streamCreator) error {
 	return <-errorChan
 }
+
+// errorDecoderV2 interprets the error channel data as plain text error message.
```
'a plain text' or 'error messages', your pick

pmorie commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -72,9 +74,9 @@ func AddSourceToErr(verb string, source string, err error) error {
 
 // BehaviorOnFatal allows you to override the default behavior when a fatal
-// error occurs, which is call os.Exit(1). You can pass 'panic' as a function
+// error occurs, which is call os.Exit(code). You can pass 'panic' as a function
```
which is to call

pmorie commented on pkg/kubelet/server/remotecommand/websocket.go:
```diff
@@ -67,9 +72,30 @@ func writeChannel(real bool) wsstream.ChannelType {
 	channels := createChannels(opts)
-	conn := wsstream.NewConn(channels...)
+	conn := wsstream.NewConn(map[string]wsstream.ChannelProtocolConfig{
+		"": {
```
what's the significance of this empty string?  comment would be helpful

sttts replied:
it's defined in the godocs o NewConn

pmorie commented on pkg/util/wsstream/conn.go:
```diff
@@ -161,7 +172,7 @@ func (conn *Conn) SetIdleTimeout(duration time.Duration) {
 }
 
 // Open the connection and create channels for reading and writing.
```
doc the return tuple here?

pmorie commented on pkg/util/wsstream/stream.go:
```diff
@@ -37,23 +38,47 @@ const binaryWebSocketProtocol = "binary.k8s.io"
 // before the stream begins reading.
-func NewReader(r io.Reader, ping bool) *Reader {
+//
+// The protocols parameter maps subprotocol names to
```
line wrapping a little weird here and elsewhere

pmorie commented on pkg/util/wsstream/stream_test.go:
```diff
@@ -69,6 +70,72 @@ func TestStreamBase64(t *testing.T) {
+func TestStreamVersionedCopy(t *testing.T) {
+	for i, test := range versionTests() {
+		func() {
+			supportedProtocols := map[string]ReaderProtocolConfig{}
```
In this test, you use `Fatalf` where I think `Errorf` would be better -- you're only going to see the first failure.  Replace with `Errorf` and `continue`

pmorie commented:
Generally LGTM, mostly looking for additional comments for those who aren't familiar with this code.  Let me know when it's ready for another pass.  Will want these commits squashed before merge.

sttts commented:
@pmorie thanks for the review :+1: I addressed all your comments in realtime. You can just restart reviewing from the beginning for a final pass ;-) Will squash at the end.

sttts commented:
@k8s-bot test this issue: #30332

sttts commented:
@pmorie @smarterclayton have switched to `unversioned.Status` for the error channel. I am a bit unsure about any necessary versioning of that object as we always are in a well-defined streaming protocol version. Moreover, do we have versioning at all on the kubelet side which provides the API?

pmorie commented:
@sttts just a note that you should squash before LGTM.

smarterclayton commented:
Unversioned is effectively "v1" for the generic server API.  The kubelet does provide versioning, but has never been rev'd.  We should doc which protocol versions are using unversioned.Status, and that should be sufficient for any client until we get to the next phase (where we have kubelet APIs with versions).

sttts commented:
@pmorie squashed
@smarterclayton I read your comment that I leave it as it is now: v4 of the streaming protocol uses `unversioned.Status`, and this is also clearly documented in the code implementing the v4 protocol.

dims commented:
LGTM 👍

smarterclayton commented on pkg/client/unversioned/remotecommand/v4.go:
```diff
@@ -0,0 +1,115 @@
+// errorDecoderV4 interprets the json-marshaled unversioned.Status on the error channel
+// and creates an exec.ExitError from it.
+type errorDecoderV4 struct{}
+
```
Is there a unit test that covers this function?  I did not see one.

sttts replied:
No there isn't, but well worth it.

sttts replied:
done

sttts commented:
@pmorie ptal

smarterclayton commented:
I have no further comments, pmorie has the conn

sttts commented:
Rebased.
